### PR TITLE
Implement HTMLDetailsElement. Fixes #9216

### LINKED ISSUE #9216: Implement HTMLDetailsElement
This requires an implementation of the `attribute_mutated` function in `VirtualMethods` for `HTMLDetailsElement`, when attempting to implement the [details notification task steps](https://html.spec.whatwg.org/multipage/#details-notification-task-steps). It also requires the usage of the DOM manipulation task source.

Code: `components/script/dom/htmldetailselement.rs` (new file), `components/script/dom/webidls/HTMLDetailsElement.webidl` (new file)
Tests: `tests/wpt/web-platform-tests/html/semantics/interactive-elements/the-details-element/details.html`
Spec: https://html.spec.whatwg.org/multipage/#htmldetailselement

Requires #9217.

### PR DESCRIPTION
**Pull request by MonsieurLanza**
Implement the interface HTMLDetailsElement ( // https://html.spec.whatwg.org/multipage/#htmldetailselement )

All tests pass in
tests/wpt/web-platform-tests/html/semantics/interactive-elements/the-details-element/details.html
&
tests/wpt/web-platform-tests/html/semantics/interactive-elements/the-details-element/toggleEvent.html

Anyway, no change is made on layout and attribute open currently has no effect, it just fires a toggle event.

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/9381)

### CODE REVIEW

**Comment by ecoal95** [comment]
It looks good to me so far (modulo those nits), but I'd like @KiChjang or @jdm to review it though.

Also, you should probably remove `tests/wpt/metadata/html/semantics/interactive-elements/the-details-element/details.html.ini` or update the test expectations.

Good job! :-)

-S-awaiting-review +S-needs-code-changes 

---

Review status: 0 of 8 files reviewed at latest revision, 2 unresolved discussions.

---

*[components/script/dom/htmldetailselement.rs, line 37 \[r1\]](https://reviewable.io:443/reviews/servo/servo/9381#-K8QLCYWdywx_cu2R5tr:-K8QLCYWdywx_cu2R5ts:-1213359478) ([raw file](https://github.com/servo/servo/blob/fdde8250ce666e3711886fd3158f46f5b36d3c3f/components/script/dom/htmldetailselement.rs#L37)):*
nit: This should be four spaces to the left

---

*[components/script/dom/htmldetailselement.rs, line 104 \[r1\]](https://reviewable.io:443/reviews/servo/servo/9381#-K8QLqTQ1nVxjdZaG1rQ:-K8QLqTQ1nVxjdZaG1rR:102728681) ([raw file](https://github.com/servo/servo/blob/fdde8250ce666e3711886fd3158f46f5b36d3c3f/components/script/dom/htmldetailselement.rs#L104)):*
nit: Indentation here could look better like:

```rust
target.upcast::<EventTarget>()
      .fire_simple_event("toggle", GlobalRef::Window(window.r()));
```

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9381#-:-K8QM99GEJyPr-O4Kzr6:-1861363292)*

**Comment by MonsieurLanza** [comment]
Oups. Forgot the detail.html.ini
Updated code to match your request.

**Comment by KiChjang** [comment]
```
Ran 4547 tests finished in 229.0 seconds.
  • 4543 ran as expected. 1369 tests skipped.
  • 4 tests had unexpected subtest results

Tests with unexpected results:
  ▶ Unexpected subtest result in /html/dom/interfaces.html:
  └ PASS [expected FAIL] Document interface: attribute ontoggle

  ▶ Unexpected subtest result in /html/dom/interfaces.html:
  └ PASS [expected FAIL] Document interface: document.implementation.createDocument(null, "", null) must inherit property "ontoggle" with the proper type (154)

  ▶ Unexpected subtest result in /html/dom/interfaces.html:
  └ PASS [expected FAIL] HTMLElement interface: attribute ontoggle

  ▶ Unexpected subtest result in /html/dom/interfaces.html:
  └ PASS [expected FAIL] HTMLElement interface: document.createElement("noscript") must inherit property "ontoggle" with the proper type (92)

  ▶ Unexpected subtest result in /html/dom/interfaces.html:
  └ PASS [expected FAIL] HTMLDetailsElement interface: existence and properties of interface object

  ▶ Unexpected subtest result in /html/dom/interfaces.html:
  └ PASS [expected FAIL] HTMLDetailsElement interface object length

  ▶ Unexpected subtest result in /html/dom/interfaces.html:
  └ PASS [expected FAIL] HTMLDetailsElement interface object name

  ▶ Unexpected subtest result in /html/dom/interfaces.html:
  └ PASS [expected FAIL] HTMLDetailsElement interface: existence and properties of interface prototype object

  ▶ Unexpected subtest result in /html/dom/interfaces.html:
  └ PASS [expected FAIL] HTMLDetailsElement interface: existence and properties of interface prototype object's "constructor" property

  ▶ Unexpected subtest result in /html/dom/interfaces.html:
  └ PASS [expected FAIL] HTMLDetailsElement interface: attribute open

  ▶ Unexpected subtest result in /html/dom/interfaces.html:
  └ PASS [expected FAIL] HTMLDetailsElement must be primary interface of document.createElement("details")

  ▶ Unexpected subtest result in /html/dom/interfaces.html:
  └ PASS [expected FAIL] Stringification of document.createElement("details")

  ▶ Unexpected subtest result in /html/dom/interfaces.html:
  └ PASS [expected FAIL] HTMLDetailsElement interface: document.createElement("details") must inherit property "open" with the proper type (0)

  ▶ Unexpected subtest result in /html/dom/reflection-misc.html:
  └ PASS [expected FAIL] details.open: typeof IDL attribute

  ▶ Unexpected subtest result in /html/dom/reflection-misc.html:
  └ PASS [expected FAIL] details.open: IDL get with DOM attribute unset

  ▶ Unexpected subtest result in /html/dom/reflection-misc.html:
  └ PASS [expected FAIL] details.open: setAttribute() to "" followed by IDL get

  ▶ Unexpected subtest result in /html/dom/reflection-misc.html:
  └ PASS [expected FAIL] details.open: setAttribute() to " foo " followed by IDL get

  ▶ Unexpected subtest result in /html/dom/reflection-misc.html:
  └ PASS [expected FAIL] details.open: setAttribute() to undefined followed by IDL get

  ▶ Unexpected subtest result in /html/dom/reflection-misc.html:
  └ PASS [expected FAIL] details.open: setAttribute() to null followed by IDL get

  ▶ Unexpected subtest result in /html/dom/reflection-misc.html:
  └ PASS [expected FAIL] details.open: setAttribute() to 7 followed by IDL get

  ▶ Unexpected subtest result in /html/dom/reflection-misc.html:
  └ PASS [expected FAIL] details.open: setAttribute() to 1.5 followed by IDL get

  ▶ Unexpected subtest result in /html/dom/reflection-misc.html:
  └ PASS [expected FAIL] details.open: setAttribute() to true followed by IDL get

  ▶ Unexpected subtest result in /html/dom/reflection-misc.html:
  └ PASS [expected FAIL] details.open: setAttribute() to false followed by IDL get

  ▶ Unexpected subtest result in /html/dom/reflection-misc.html:
  └ PASS [expected FAIL] details.open: setAttribute() to object "[object Object]" followed by IDL get

  ▶ Unexpected subtest result in /html/dom/reflection-misc.html:
  └ PASS [expected FAIL] details.open: setAttribute() to NaN followed by IDL get

  ▶ Unexpected subtest result in /html/dom/reflection-misc.html:
  └ PASS [expected FAIL] details.open: setAttribute() to Infinity followed by IDL get

  ▶ Unexpected subtest result in /html/dom/reflection-misc.html:
  └ PASS [expected FAIL] details.open: setAttribute() to -Infinity followed by IDL get

  ▶ Unexpected subtest result in /html/dom/reflection-misc.html:
  └ PASS [expected FAIL] details.open: setAttribute() to "\\0" followed by IDL get

  ▶ Unexpected subtest result in /html/dom/reflection-misc.html:
  └ PASS [expected FAIL] details.open: setAttribute() to object "test-toString" followed by IDL get

  ▶ Unexpected subtest result in /html/dom/reflection-misc.html:
  └ PASS [expected FAIL] details.open: setAttribute() to object "test-valueOf" followed by IDL get

  ▶ Unexpected subtest result in /html/dom/reflection-misc.html:
  └ PASS [expected FAIL] details.open: setAttribute() to "open" followed by IDL get

  ▶ Unexpected subtest result in /html/dom/reflection-misc.html:
  └ PASS [expected FAIL] details.open: IDL set to "" followed by hasAttribute()

  ▶ Unexpected subtest result in /html/dom/reflection-misc.html:
  └ PASS [expected FAIL] details.open: IDL set to "" followed by IDL get

  ▶ Unexpected subtest result in /html/dom/reflection-misc.html:
  └ PASS [expected FAIL] details.open: IDL set to " foo " followed by IDL get

  ▶ Unexpected subtest result in /html/dom/reflection-misc.html:
  └ PASS [expected FAIL] details.open: IDL set to undefined followed by hasAttribute()

  ▶ Unexpected subtest result in /html/dom/reflection-misc.html:
  └ PASS [expected FAIL] details.open: IDL set to undefined followed by IDL get

  ▶ Unexpected subtest result in /html/dom/reflection-misc.html:
  └ PASS [expected FAIL] details.open: IDL set to null followed by hasAttribute()

  ▶ Unexpected subtest result in /html/dom/reflection-misc.html:
  └ PASS [expected FAIL] details.open: IDL set to null followed by IDL get

  ▶ Unexpected subtest result in /html/dom/reflection-misc.html:
  └ PASS [expected FAIL] details.open: IDL set to 7 followed by IDL get

  ▶ Unexpected subtest result in /html/dom/reflection-misc.html:
  └ PASS [expected FAIL] details.open: IDL set to 1.5 followed by IDL get

  ▶ Unexpected subtest result in /html/dom/reflection-misc.html:
  └ PASS [expected FAIL] details.open: IDL set to false followed by hasAttribute()

  ▶ Unexpected subtest result in /html/dom/reflection-misc.html:
  └ PASS [expected FAIL] details.open: IDL set to object "[object Object]" followed by IDL get

  ▶ Unexpected subtest result in /html/dom/reflection-misc.html:
  └ PASS [expected FAIL] details.open: IDL set to NaN followed by hasAttribute()

  ▶ Unexpected subtest result in /html/dom/reflection-misc.html:
  └ PASS [expected FAIL] details.open: IDL set to NaN followed by IDL get

  ▶ Unexpected subtest result in /html/dom/reflection-misc.html:
  └ PASS [expected FAIL] details.open: IDL set to Infinity followed by IDL get

  ▶ Unexpected subtest result in /html/dom/reflection-misc.html:
  └ PASS [expected FAIL] details.open: IDL set to -Infinity followed by IDL get

  ▶ Unexpected subtest result in /html/dom/reflection-misc.html:
  └ PASS [expected FAIL] details.open: IDL set to "\\0" followed by IDL get

  ▶ Unexpected subtest result in /html/dom/reflection-misc.html:
  └ PASS [expected FAIL] details.open: IDL set to object "test-toString" followed by IDL get

  ▶ Unexpected subtest result in /html/dom/reflection-misc.html:
  └ PASS [expected FAIL] details.open: IDL set to object "test-valueOf" followed by IDL get

  ▶ Unexpected subtest result in /html/semantics/interfaces.html:
  └ PASS [expected FAIL] Interfaces for details

  ▶ Unexpected subtest result in /html/semantics/interfaces.html:
  └ PASS [expected FAIL] Interfaces for DETAILS

  ▶ Unexpected subtest result in /_mozilla/mozilla/interfaces.html:
  │ FAIL [expected PASS] Interfaces exposed on the window
  │   → assert_true: If this is failing: DANGER, are you sure you want to expose the new interface HTMLDetailsElement to all webpages as a property on the window? Do not make a change to this file without review from jdm or Ms2ger for that specific change! expected true got false
  │ 
  │ @http://web-platform.test:8000/_mozilla/mozilla/interfaces.html:252:1
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1381:20
  │ test@http://web-platform.test:8000/resources/testharness.js:495:9
  └ @http://web-platform.test:8000/_mozilla/mozilla/interfaces.html:246:1
```

This is looking good!

**Comment by KiChjang** [comment]
In particular though, you'll need to add `HTMLDetailsElement` to the `interfaceNamesInGlobalScope` variable in `tests/wpt/mozilla/tests/mozilla/interfaces.html`

**Comment by MonsieurLanza** [comment]
I just saw this, yes. But it somewhat says I need God's approval to do so.

**Comment by KiChjang** [comment]
@MonsieurLanza The Divine @jdm has vested powers into me and I approve of you to add in `HTMLDetailsElement` to the list of `interfaceNamesInGlobalScope`.

**Comment by ecoal95** [comment]
Yay, those failures look legit \o/

File: tests/wpt/metadata/html/dom/interfaces.html.ini [hunk 1]
```diff
@@ -216,10 +216,7 @@
   [Document interface: attribute ontimeupdate]
     expected: FAIL
 
-  [Document interface: attribute ontoggle]
-    expected: FAIL
-
-  [Document interface: attribute onvolumechange]
+ [Document interface: attribute onvolumechange]
```
Comment (KiChjang) [inline tests/wpt/metadata/html/dom/interfaces.html.ini]: A space is removed here. I think this is unintended. Please revert this.

**Comment by MonsieurLanza** [comment]
Reverted that pagan space. ^^

**Comment by KiChjang** [comment]
@bors-servo r+

Thanks for working on this!

**Pull request merged**